plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/async/AsyncUploadUtils.java

```diff
@@ -0,0 +1,460 @@
+            } else {
+                log.debug(() -> "Starting the upload as multipart upload request");
+                uploadInParts(s3AsyncClient, uploadRequest, streamContext, returnFuture);
+            }
+        } catch (Throwable throwable) {
+            returnFuture.completeExceptionally(throwable);
+        }
+
+        return returnFuture;
+    }
+
+    private void uploadInParts(
+        S3AsyncClient s3AsyncClient,
+        UploadRequest uploadRequest,
+        StreamContext streamContext,
+        CompletableFuture<Void> returnFuture
+    ) {
+
+        CreateMultipartUploadRequest.Builder createMultipartUploadRequestBuilder = CreateMultipartUploadRequest.builder()
+            .bucket(uploadRequest.getBucket())
+            .key(uploadRequest.getKey());
+        if (uploadRequest.doRemoteDataIntegrityCheck()) {
+            createMultipartUploadRequestBuilder.checksumAlgorithm(ChecksumAlgorithm.CRC32);
+        }
+        CompletableFuture<CreateMultipartUploadResponse> createMultipartUploadFuture = SocketAccess.doPrivileged(
+            () -> s3AsyncClient.createMultipartUpload(createMultipartUploadRequestBuilder.build())
+        );
+
+        // Ensure cancellations are forwarded to the createMultipartUploadFuture future
+        CompletableFutureUtils.forwardExceptionTo(returnFuture, createMultipartUploadFuture);
+
+        createMultipartUploadFuture.whenComplete((createMultipartUploadResponse, throwable) -> {
+            if (throwable != null) {
+                handleException(returnFuture, () -> "Failed to initiate multipart upload", throwable);
+            } else {
+                log.debug(() -> "Initiated new multipart upload, uploadId: " + createMultipartUploadResponse.uploadId());
+                doUploadInParts(s3AsyncClient, uploadRequest, streamContext, returnFuture, createMultipartUploadResponse.uploadId());
+            }
+        });
+    }
+
+    private void doUploadInParts(
+        S3AsyncClient s3AsyncClient,
+        UploadRequest uploadRequest,
+        StreamContext streamContext,
+        CompletableFuture<Void> returnFuture,
+        String uploadId
+    ) {
+
+        // The list of completed parts must be sorted
+        AtomicReferenceArray<CompletedPart> completedParts = new AtomicReferenceArray<>(streamContext.getNumberOfParts());
+        AtomicReferenceArray<InputStreamCRC32Container> inputStreamContainers = new AtomicReferenceArray<>(
+            streamContext.getNumberOfParts()
+        );
+
+        List<CompletableFuture<CompletedPart>> futures;
+        try {
+            futures = sendUploadPartRequests(s3AsyncClient, uploadRequest, streamContext, uploadId, completedParts, inputStreamContainers);
+        } catch (Exception ex) {
+            try {
+                cleanUpParts(s3AsyncClient, uploadRequest, uploadId);
+            } finally {
+                returnFuture.completeExceptionally(ex);
```

Comment:
I believe you are referring to https://github.com/opensearch-project/OpenSearch/pull/7218/files#diff-a472eda8bf5f051224172440b745603502d3545f8326b28d6014a79d5a833cd5R165.

The purpose of this line is to create an aggregated single future that would be completed normally/exceptionally depending on how all the futures from the `futures` list are completed. If all futures are completed normally, the resultant future is also completed normally. If any of them is completed exceptionally, the resultant future will also be completed exceptionally.

Reply:
Execution of a future starts when a `get` or a `join` is invoked on a future. We only create future objects against each part request till this point. Execution of these futures is not triggered here. Having said that this part of code is executed when we get a successful response of create multipart upload request check #L129. What this means at this point of code is that there are inactive future objects w.r.t to parts present and there is one active (completed) future object of create multipart upload present. In case of an exception during creation of a part future object, the only thing which needs to be done is to abort the request. Now, returnFuture is the one on which caller is expecting something (request/exception). That is why we are invoking `completeExceptionally` on this future to notify the caller.